torchao/prototype/autoround/core.py

```diff
@@ -0,0 +1,269 @@
+                    quant_min=quant_min,
+                    quant_max=quant_max,
+                    zero_point_domain=ZeroPointDomain.FLOAT,
+                    layout_type=TensorCoreTiledLayoutType(inner_k_tiles=inner_k_tiles),
+                )
+
+            # TODO(Yi): better way to select the weight quantization function
+            if (
+                _auto_round_config.bits == 4
+                and observed_linear.weight.device.type == "cuda"
+            ):
+                weight_func = to_int4_tinygemm_weight
+            else:
+                weight_func = to_uintx_weight
+
+            observed_linear.weight = torch.nn.Parameter(
+                weight_func(observed_linear.weight), requires_grad=False
+            )
+            del observed_linear.scale
+            del observed_linear.zp
+            return observed_linear
+
+        def _is_observed_linear(mod: torch.nn.Module, fqn: str):
+            return hasattr(mod, "scale")
+
+        qmodel = ao_quant.quant_api._replace_with_custom_fn_if_matches_filter(
+            optimized_model,
+            convert_weight_to_affine_quantized_tensor,
+            _is_observed_linear,
+        )
+        return qmodel
+
+    return _apply_auto_round
+
+
+@torch.no_grad()
+def _apply_auto_round_optimization(
+    block, grouped_args, spec, block_outputs, config: _AutoRoundConfig
+):
+    # Call the auto-round to execute the optimization process.
+    # https://github.com/intel/auto-round/tree/patch-for-ao-2
+    # TODO(Yi), make the branch more stable
+    if ar_utils.is_auto_round_available():
+        import auto_round
+    else:
+        raise ImportError(
+            (
+                "This example requires the `auto-round` library."
+                "Please install it with `pip install https://github.com/intel/auto-round.git@patch-for-ao-2`"
+            )
+        )
+    block = block.to(_multi_tensor_config.accelerator_name)
+    _optimization_tracker.optimized_layers += 1
+    logging.warning(
+        "Apply auto-round optimization on layer %d / %d.",
+        _optimization_tracker.optimized_layers,
+        _optimization_tracker.num_layers,
+    )
+
+    # Start the training process to update the v, alpha and betta.
+    rounder = auto_round.AutoRound(
+        model=block,
+        tokenizer=None,
+        sym=False,  # Both `True` and `False` are OK, but use `asym` by default for using the `tinygemm` by default
+        bits=config.bits,
+        iters=config.iters,
+        group_size=config.group_size,
+        use_quant_input=False,  # disable it for now
+        amp=True,
+        model_dtype=next(block.parameters()).dtype,
+    )
+
+    @torch.no_grad()
+    def _unflatten_grouped_args(grouped_args, spec):
+        inputs = []
+        for inp in grouped_args:
+            cur_args, cur_kwargs = tree_unflatten(inp, spec)
+            inputs.append((cur_args, cur_kwargs))
+        return inputs
+
+    block_inputs = _unflatten_grouped_args(grouped_args, spec)
```

Comment:
I'm wondering if these things can be simplified as well, these seems to be relying on a lot of internal details

Reply:
Yeah, I updated it to use `revert_to_tensor_pairs` and add more explanation at https://github.com/pytorch/ao/pull/581/files#r1730624604